src/openms/include/OpenMS/METADATA/IdentificationData.h

```diff
@@ -0,0 +1,320 @@
+      }
+    };
+
+    typedef UInt64 ScoreTypeKey;
+    typedef boost::bimap<ScoreTypeKey, ScoreType> ScoreTypeBimap;
+    ScoreTypeBimap score_types;
+
+
+    /*!
+      Search query, e.g. spectrum or feature.
+    */
+    struct DataQuery: public MetaInfoInterface
+    {
+      InputFileKey input_file_key; // reference into "input_files"
+
+      // spectrum or feature ID (from the file reference by "input_file_key"):
+      String data_id;
+
+      double rt, mz; // position
+
+      // @TODO: ignore RT and m/z for comparisons?
+      bool operator<(const DataQuery& other) const
+        {
+          return std::tie(input_file_key, data_id, rt, mz) <
+            std::tie(other.input_file_key, other.data_id, other.rt, other.mz);
+        }
+
+      // @TODO: ignore RT and m/z for comparisons?
+      bool operator==(const DataQuery& other) const
+        {
+          return std::tie(input_file_key, data_id, rt, mz) ==
+            std::tie(other.input_file_key, other.data_id, other.rt, other.mz);
+        }
+    };
+
+    typedef UInt64 DataQueryKey;
+    typedef boost::bimap<DataQueryKey, DataQuery> QueryBimap;
+    QueryBimap data_queries;
+
+
+    // Identified molecules - at the moment, peptides or small molecules:
+    typedef UInt64 IdentifiedMoleculeKey;
+    typedef boost::bimap<IdentifiedMoleculeKey, AASequence> PeptideBimap;
+    typedef boost::bimap<IdentifiedMoleculeKey, String> CompoundBimap;
+    PeptideBimap identified_peptides;
+    CompoundBimap identified_compounds;
+
+    enum MoleculeType
+    {
+      MT_PROTEIN,
+      MT_COMPOUND,
+      // MT_RNA,
+      SIZE_OF_MOLECULETYPES
+    };
+
+    /*!
+      Meta data for an identified molecule.
+    */
+    struct IdentifiedMetaData: public MetaInfoInterface
+    {
+      enum MoleculeType molecule_type;
+
+      std::unordered_map<ScoreTypeKey, double> scores;
+
+      std::vector<ProcessingStepKey> processing_steps;
+    };
+
+    std::unordered_map<IdentifiedMoleculeKey,
+                       IdentifiedMetaData> identified_meta_data;
+
+
+    // Evidence linking identified molecules to parent molecules (e.g. peptides
+    // to proteins):
+    // @TODO: rename "PeptideEvidence" to "MoleculeEvidence" (incl. members)
+    typedef std::unordered_map<IdentifiedMoleculeKey,
+                               std::vector<PeptideEvidence>> EvidenceMap;
+    EvidenceMap parent_evidence;
+
+
+    /*!
+      Meta data specific to an identified compound (small molecule).
+    */
+    struct CompoundMetaData: public MetaInfoInterface
+    {
+      EmpiricalFormula formula;
+
+      String smile;
+
+      String inchi;
+    };
+
+    std::unordered_map<IdentifiedMoleculeKey,
+                       CompoundMetaData> compound_meta_data;
+
+
+    /*!
+      Meta data for a search hit (e.g. peptide-spectrum match).
+    */
+    struct MatchMetaData: public MetaInfoInterface
+    {
+      Int charge;
+
+      std::unordered_map<ScoreTypeKey, double> scores;
```

Comment:
probably need a way to define primary score

Reply:
This is an important point, but unfortunately it's - probably? - not enough to just designate one score type as the primary one (since you may have different types of molecules, e.g. peptides vs. compounds, and different levels, e.g. PSMs, peptides, proteins). You could store a different primary score for each molecule/level/..., but maybe there's a better solution? I haven't thought it through yet.

Reply:
I think I've found a good (and quite simple) solution: Replace all the `unordered_map<ScoreTypeKey, double>` with `vector<pair<ScoreTypeKey, double>>`. Then the order in the vector would give a precedence, with the last entry being the "primary score". This would work quite naturally as sequential processing steps append their scores to the list, so that new scores would take precedence over old ones.
Trade-offs:
- less efficient look-up of scores by score type (but the list shouldn't get very long anyway),
- no guarantee that entries are unique by score type (but that doesn't seem very important and we'd still have a clearly defined order of precedence).